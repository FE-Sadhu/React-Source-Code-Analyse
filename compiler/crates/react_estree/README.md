# react_estree

This crate is a Rust representation of the [ESTree format](https://github.com/estree/estree/tree/master) and
popular extenions including JSX and (eventually) Flow and TypeScript.

This crate is intended as the main interchange format with outside code. A typical integration with React Compiler
will look as follows:

1. Host Compiler parses into the host AST format.
2. Host Compiler converts into `react_estree`.
3. Host Compiler invokes React Compiler to compile the input, which (conceptually) 
   returns the resulting code in `react_estree` format.
4. Host Compiler convert back from `react_estree` to its host AST format.

Because React Compiler is intended to support JavaScript-based toolchains, `react_estree` is designed to support
accurate serialization to/from estree-compatible JSON. We may also support the Babel AST format 
(a variant of ESTree) as well, depending on demand.