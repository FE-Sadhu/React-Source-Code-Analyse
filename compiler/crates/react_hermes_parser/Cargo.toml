[package]
name = "react_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
react_diagnostics = { workspace = true }
react_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
react_estree_codegen = { workspace = true }