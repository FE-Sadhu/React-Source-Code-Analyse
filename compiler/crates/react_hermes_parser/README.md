# react_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `react_estree` AST.