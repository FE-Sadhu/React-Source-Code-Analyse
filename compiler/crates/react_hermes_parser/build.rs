/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use react_estree_codegen::estree_hermes;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../react_estree_codegen");

    let src = estree_hermes();
    let copyright = "
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
    "
    .to_string();
    let trimmed_copyright = copyright.trim();
    let contents = format!("{trimmed_copyright}\n{src}");
    std::fs::write("src/generated.rs", contents).unwrap();
}
