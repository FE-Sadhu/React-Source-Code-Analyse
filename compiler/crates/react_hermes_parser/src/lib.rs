/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

mod generated;
mod generated_extension;

use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;
use react_diagnostics::Diagnostic;
use react_estree::Program;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new(&buf);
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
