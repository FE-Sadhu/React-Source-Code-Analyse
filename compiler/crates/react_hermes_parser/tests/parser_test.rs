/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::env;

use insta::{assert_snapshot, glob};
use react_estree::SourceType;
use react_hermes_parser::parse;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let mut ast = parse(&input, path.to_str().unwrap()).unwrap();
        // TODO: hack to prevent changing lots of fixtures all at once
        ast.source_type = SourceType::Script;
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
