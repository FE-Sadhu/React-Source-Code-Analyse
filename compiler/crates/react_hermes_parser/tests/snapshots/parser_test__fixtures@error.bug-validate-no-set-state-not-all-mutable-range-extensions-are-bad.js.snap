---
source: crates/react_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/react_hermes_parser/tests/fixtures/error.bug-validate-no-set-state-not-all-mutable-range-extensions-are-bad.js
---
Input:
function Component(props) {
  const logEvent = useLogging(props.appId);
  const [currentStep, setCurrentStep] = useState(0);

  const onSubmit = (errorEvent) => {
    // 2. onSubmit inherits the mutable range of logEvent
    logEvent(errorEvent);
    // 3. this call then triggers the ValidateNoSetStateInRender check incorrectly, even though
    //    onSubmit is not called during render (although it _could_ be, if OtherComponent does so.
    //    but we can't tell without x-file analysis)
    setCurrentStep(1);
  };

  switch (currentStep) {
    case 0:
      return <OtherComponent data={{ foo: "bar" }} />;
    case 1:
      return <OtherComponent data={{ foo: "joe" }} onSubmit={onSubmit} />;
    default:
      // 1. logEvent's mutable range is extended to this instruction
      logEvent("Invalid step");
      return <OtherComponent data={null} />;
  }
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          9,
          18
        ]
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": [
            19,
            24
          ]
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "logEvent",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": [
                    36,
                    44
                  ]
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useLogging",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      47,
                      57
                    ]
                  },
                  "arguments": [
                    {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          58,
                          63
                        ]
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "appId",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          64,
                          69
                        ]
                      },
                      "computed": false,
                      "loc": null,
                      "range": [
                        58,
                        69
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    47,
                    70
                  ]
                },
                "loc": null,
                "range": [
                  36,
                  70
                ]
              }
            ],
            "loc": null,
            "range": [
              30,
              71
            ]
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "currentStep",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        81,
                        92
                      ]
                    },
                    {
                      "type": "Identifier",
                      "name": "setCurrentStep",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        94,
                        108
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    80,
                    109
                  ]
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useState",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      112,
                      120
                    ]
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 0.0,
                      "loc": null,
                      "range": [
                        121,
                        122
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    112,
                    123
                  ]
                },
                "loc": null,
                "range": [
                  80,
                  123
                ]
              }
            ],
            "loc": null,
            "range": [
              74,
              124
            ]
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "onSubmit",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": [
                    134,
                    142
                  ]
                },
                "init": {
                  "type": "ArrowFunctionExpression",
                  "id": null,
                  "params": [
                    {
                      "type": "Identifier",
                      "name": "errorEvent",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        146,
                        156
                      ]
                    }
                  ],
                  "body": {
                    "type": "BlockStatement",
                    "body": [
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "CallExpression",
                          "callee": {
                            "type": "Identifier",
                            "name": "logEvent",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": [
                              225,
                              233
                            ]
                          },
                          "arguments": [
                            {
                              "type": "Identifier",
                              "name": "errorEvent",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": [
                                234,
                                244
                              ]
                            }
                          ],
                          "loc": null,
                          "range": [
                            225,
                            245
                          ]
                        },
                        "directive": null,
                        "loc": null,
                        "range": [
                          225,
                          246
                        ]
                      },
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "CallExpression",
                          "callee": {
                            "type": "Identifier",
                            "name": "setCurrentStep",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": [
                              499,
                              513
                            ]
                          },
                          "arguments": [
                            {
                              "type": "NumericLiteral",
                              "value": 1.0,
                              "loc": null,
                              "range": [
                                514,
                                515
                              ]
                            }
                          ],
                          "loc": null,
                          "range": [
                            499,
                            516
                          ]
                        },
                        "directive": null,
                        "loc": null,
                        "range": [
                          499,
                          517
                        ]
                      }
                    ],
                    "loc": null,
                    "range": [
                      161,
                      521
                    ]
                  },
                  "generator": false,
                  "async": false,
                  "loc": null,
                  "range": [
                    145,
                    521
                  ],
                  "expression": false,
                  "loc": null,
                  "range": [
                    145,
                    521
                  ]
                },
                "loc": null,
                "range": [
                  134,
                  521
                ]
              }
            ],
            "loc": null,
            "range": [
              128,
              522
            ]
          },
          {
            "type": "SwitchStatement",
            "discriminant": {
              "type": "Identifier",
              "name": "currentStep",
              "typeAnnotation": null,
              "loc": null,
              "range": [
                534,
                545
              ]
            },
            "cases": [
              {
                "type": "SwitchCase",
                "test": {
                  "type": "NumericLiteral",
                  "value": 0.0,
                  "loc": null,
                  "range": [
                    558,
                    559
                  ]
                },
                "consequent": [
                  {
                    "type": "ReturnStatement",
                    "argument": {
                      "type": "JSXElement",
                      "openingElement": {
                        "type": "JSXOpeningElement",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "OtherComponent",
                          "loc": null,
                          "range": [
                            575,
                            589
                          ]
                        },
                        "attributes": [
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "data",
                              "loc": null,
                              "range": [
                                590,
                                594
                              ]
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "ObjectExpression",
                                "properties": [
                                  {
                                    "type": "Property",
                                    "key": {
                                      "type": "Identifier",
                                      "name": "foo",
                                      "typeAnnotation": null,
                                      "loc": null,
                                      "range": [
                                        598,
                                        601
                                      ]
                                    },
                                    "value": {
                                      "type": "StringLiteral",
                                      "value": "bar",
                                      "loc": null,
                                      "range": [
                                        603,
                                        608
                                      ]
                                    },
                                    "kind": "init",
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "loc": null,
                                    "range": [
                                      598,
                                      608
                                    ]
                                  }
                                ],
                                "loc": null,
                                "range": [
                                  596,
                                  610
                                ]
                              },
                              "loc": null,
                              "range": [
                                595,
                                611
                              ]
                            },
                            "loc": null,
                            "range": [
                              590,
                              611
                            ]
                          }
                        ],
                        "selfClosing": true,
                        "loc": null,
                        "range": [
                          574,
                          614
                        ]
                      },
                      "children": [],
                      "closingElement": null,
                      "loc": null,
                      "range": [
                        574,
                        614
                      ]
                    },
                    "loc": null,
                    "range": [
                      567,
                      615
                    ]
                  }
                ],
                "loc": null,
                "range": [
                  553,
                  615
                ]
              },
              {
                "type": "SwitchCase",
                "test": {
                  "type": "NumericLiteral",
                  "value": 1.0,
                  "loc": null,
                  "range": [
                    625,
                    626
                  ]
                },
                "consequent": [
                  {
                    "type": "ReturnStatement",
                    "argument": {
                      "type": "JSXElement",
                      "openingElement": {
                        "type": "JSXOpeningElement",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "OtherComponent",
                          "loc": null,
                          "range": [
                            642,
                            656
                          ]
                        },
                        "attributes": [
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "data",
                              "loc": null,
                              "range": [
                                657,
                                661
                              ]
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "ObjectExpression",
                                "properties": [
                                  {
                                    "type": "Property",
                                    "key": {
                                      "type": "Identifier",
                                      "name": "foo",
                                      "typeAnnotation": null,
                                      "loc": null,
                                      "range": [
                                        665,
                                        668
                                      ]
                                    },
                                    "value": {
                                      "type": "StringLiteral",
                                      "value": "joe",
                                      "loc": null,
                                      "range": [
                                        670,
                                        675
                                      ]
                                    },
                                    "kind": "init",
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "loc": null,
                                    "range": [
                                      665,
                                      675
                                    ]
                                  }
                                ],
                                "loc": null,
                                "range": [
                                  663,
                                  677
                                ]
                              },
                              "loc": null,
                              "range": [
                                662,
                                678
                              ]
                            },
                            "loc": null,
                            "range": [
                              657,
                              678
                            ]
                          },
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "onSubmit",
                              "loc": null,
                              "range": [
                                679,
                                687
                              ]
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "Identifier",
                                "name": "onSubmit",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": [
                                  689,
                                  697
                                ]
                              },
                              "loc": null,
                              "range": [
                                688,
                                698
                              ]
                            },
                            "loc": null,
                            "range": [
                              679,
                              698
                            ]
                          }
                        ],
                        "selfClosing": true,
                        "loc": null,
                        "range": [
                          641,
                          701
                        ]
                      },
                      "children": [],
                      "closingElement": null,
                      "loc": null,
                      "range": [
                        641,
                        701
                      ]
                    },
                    "loc": null,
                    "range": [
                      634,
                      702
                    ]
                  }
                ],
                "loc": null,
                "range": [
                  620,
                  702
                ]
              },
              {
                "type": "SwitchCase",
                "test": null,
                "consequent": [
                  {
                    "type": "ExpressionStatement",
                    "expression": {
                      "type": "CallExpression",
                      "callee": {
                        "type": "Identifier",
                        "name": "logEvent",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          791,
                          799
                        ]
                      },
                      "arguments": [
                        {
                          "type": "StringLiteral",
                          "value": "Invalid step",
                          "loc": null,
                          "range": [
                            800,
                            814
                          ]
                        }
                      ],
                      "loc": null,
                      "range": [
                        791,
                        815
                      ]
                    },
                    "directive": null,
                    "loc": null,
                    "range": [
                      791,
                      816
                    ]
                  },
                  {
                    "type": "ReturnStatement",
                    "argument": {
                      "type": "JSXElement",
                      "openingElement": {
                        "type": "JSXOpeningElement",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "OtherComponent",
                          "loc": null,
                          "range": [
                            831,
                            845
                          ]
                        },
                        "attributes": [
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "data",
                              "loc": null,
                              "range": [
                                846,
                                850
                              ]
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "NullLiteral",
                                "loc": null,
                                "range": [
                                  852,
                                  856
                                ]
                              },
                              "loc": null,
                              "range": [
                                851,
                                857
                              ]
                            },
                            "loc": null,
                            "range": [
                              846,
                              857
                            ]
                          }
                        ],
                        "selfClosing": true,
                        "loc": null,
                        "range": [
                          830,
                          860
                        ]
                      },
                      "children": [],
                      "closingElement": null,
                      "loc": null,
                      "range": [
                        830,
                        860
                      ]
                    },
                    "loc": null,
                    "range": [
                      823,
                      861
                    ]
                  }
                ],
                "loc": null,
                "range": [
                  707,
                  861
                ]
              }
            ],
            "loc": null,
            "range": [
              526,
              865
            ]
          }
        ],
        "loc": null,
        "range": [
          26,
          867
        ]
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": [
        0,
        867
      ],
      "loc": null,
      "range": [
        0,
        867
      ]
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": [
    0,
    867
  ]
}
